<?xml version="1.0"?>
<launch>

	<!-- arg robot -->
  	<arg name="robot" default="-1" doc="If robot is 0, turtlebot runs the code itself, else, code is run remotely from the computer"/>
		<arg name="world_name" default="$(find nuturtlebot)/worlds/block.world"/>

	<!--Robot Description from URDF-->
  <rosparam file="$(find nuturtle_description)/config/diff_params.yaml" command="load"/>

	<!--Include basic_remote-->
	<include file="$(find nuturtle_robot)/launch/basic_remote.launch">
	  	<arg name="robot" value="$(arg robot)" />
	</include>

	<!--node turtle_interface-->
	<node machine="computer" name="turtle_interface" pkg="nuturtle_robot" type="turtle_interface"/>

	<node machine="computer" name="teleop_twist_keyboard" pkg="turtlebot3_teleop" type="turtlebot3_teleop_key"/>

	<!--node fake_diff_encoders-->
	<node machine="computer" name="fake_diff_encoders" pkg="rigid2d" type="fake_diff_encoders">
		<remap from="/turtle1/cmd_vel" to="/cmd_vel"/>			
		<remap from="joint_states" to="fake/joint_states"/>
	</node>
	
	<!--odometry for the real robot-->
	<node machine="computer" name="odometer" pkg="rigid2d" type="odometer">
		<param name="odom_frame_id" type="str" value="odom" />
		<param name="body_frame_id" type="str" value="base_link" />
	</node>

	<!--odometry from the fake encoder-->
	<node machine="computer" name="odometer2" pkg="rigid2d" type="odometer" output="screen" ns="fake">
		<param name="odom_frame_id" type="str" value="/fake_odom" />
		<param name="body_frame_id" type="str" value="/fake_base_link" />
		<remap from="joint_states" to="joint_states"/>
		<remap from="odom_pub" to="odom"/>
	</node>

	<!--bind odom with fake_odom-->
	<node pkg="tf2_ros" type="static_transform_publisher" name="tf_trans" args="0 0 0 0 0 0 1 odom fake_odom" />

	<!--robot_state_publisher-->
	<node machine="computer" name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>
	
	<!--enable rviz-->
	<group unless="$(eval arg('robot') ==0)">
		<node name="rviz" pkg="rviz" type="rviz" args="-d $(find nuturtle_robot)/config/config.rviz" />
	</group>

	<!-- ros_args.launch -->
	<group if="$(eval arg('robot') ==-1)">
		<include file="$(find nuslam)/launch/diff_drive_gazebo.launch"/>
	</group>

</launch>
